src/Particular.HealthMonitoring.Uptime/UptimeMonitoring.cs

```diff
@@ -1,51 +1,41 @@
 ﻿namespace Particular.HealthMonitoring.Uptime
 {
-    using System;
-    using System.Threading;
-    using ServiceControl.Infrastructure;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Particular.HealthMonitoring.Uptime.Api;
     using ServiceControl.Infrastructure.DomainEvents;
-    using ServiceControl.Monitoring;
 
-    public class UptimeMonitoring
+    public class UptimeMonitoringDependencies
     {
-        public EndpointInstanceMonitoring Monitoring { get; }
+        public IPersistEndpointUptimeInformation Persister { get; set; }
+        public IDomainEvents DomainEvents { get; set; }
+    }
 
-        ITimeKeeper timeKeeper;
-        TimeSpan gracePeriod;
-        Timer timer;
+    public class UptimeMonitoring : IComponent<UptimeMonitoringDependencies>
+    {
+        EndpointInstanceMonitoring monitoring;
 
-        public UptimeMonitoring(TimeSpan gracePeriod, IDomainEvents domainEvents, ITimeKeeper timeKeeper)
+        public UptimeMonitoring()
         {
-            this.timeKeeper = timeKeeper;
-            this.gracePeriod = gracePeriod;
-
-            Monitoring = new EndpointInstanceMonitoring(domainEvents);
+            monitoring = new EndpointInstanceMonitoring();
         }
-
-        public void Start()
+        
+        public IEnumerable<object> CreateParts()
```

Comment:
What is a `Part`? These seem a little like a piece of a `Component`. Is there a better word somewhere between "thing" and "component"?

Reply:
Yeah. Part is not the best one. Maybe `component` -> `composite` & `part` -> `component`?

Reply:
I would propose that we create a ComponentInitializationContext or something that knows how to handle these things so that we can handle them differently depending on what they are. Envisaging something like this:

```csharp
    async Task Initialize(ComponentInitContext context)
    {
        context.Services.Register<EndpointInstanceMonitoring>();
        context.Services.RegisterDefault<IHeartbeatPersistence, NullHeartbeatPersistence>();

        context.Jobs.Register<MonitorEndpointInstances>();

        context.Handlers.Register<EndpointHeartbeatHandler>();
        context.Handlers.Register<EndpointStartedHandler>();

        context.Api.Get<HeartbeatStatsApi>("/heartbeats/stats");
        context.Api.Get<MonitoringEndpointsApi>("/endpoints");
        context.Api.Patch<ToggleMonitoredEndpointsApi>("/endpoints/{id}");
    }
```

Reply:
The `ComponentInitContext` approach would require that all possible APIs live in the same package and all components depend on all APIs. Since these are only APIs I guess the danger of versioning issues is not that big.

On the other hand I don't see added value. Comparing:

```
context.RegisterPart<EndpointStartedHandler>();
```

versus

```
context.Handlers.Register<EndpointStartedHandler>();
```

Another issue is lifecycle management. I don't think we need a container. We're moving away from container usage in `Features` of the core because of complexity. The model in the core is following:
 * Feature (root, equivalent of `IComponent` here) creates shared objects e.g. `EndpointInstanceMonitoring`)
 * Feature creates various things (parts here) e.g. behaviors as instances and registers them

I think this simple model should be enough for our purposes here as well.